### DISCUSSION TITLE
Persist data with Next JS (pages)

### DISCUSSION THREAD

**Discussion by nikitassh** [vercel/next.js#56894] [Help]
### Summary

Hi all

please tell me if I have a login page by phone number and a phone number confirmation page

I need to transfer a phone number from page A to page B, what is the best way to do this?

I use zustand

**Comment by icyJoseph** [comment]
Annoyingly, the easiest way is to hoist the data to a Context Provider on `_app`, and consume it from there.



```tsx
const TelephoneProvider = ({children}) => {
    const [telephone, setTelephone] = useState("");

    return (
       <TelephoneNumberContext.Provider value={telephone}>
            <TelephoneNumberSetterContext.Provider value={setTelephone}>
                {children}
            </TelephoneNumberSetterContext.Provider>
       </TelephoneNumberContext.Provider>
    )
}

// export functions to useContext(TelephoneNumberSetterContext) and useContext(TelephoneNumberContext)
```

And then wrap around the return of `_app`. This way components that set the telephone number won't necessarily re-render when it is set, but that's an optimization you may not need just yet.

Does this mean that all pages get the ability to possibly, access this context? Yes... but this is one way.

Another, would be to use URL queries, but that might open up for people trying to by pass, or inject phone numbers, etc..

**Reply by nikitassh** [reply]
But after the user refreshes the page, my context will be updated and the phone will become empty again

**Reply by icyJoseph** [reply]
Right, and for that you can use something like `localStorage`. As you set the telephone number, you can save it to localStorage, and on page refresh, you can synchronize with the `localStorage` inside a use-effect. Not during the render pass.

- https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage
